### PR TITLE
DISTX-266

### PR DESCRIPTION
**Pull request by handavid**
**NOTE: the first two commits are from https://github.com/hortonworks/cloudbreak/pull/5653.
I included them here because I move one of the files and would have a rebase conflict without them.**

The FMS was enhanced to better track SyncOperations and reject
concurrent requests.
    
The syncoperation table was modified
to add the environmentList and userList columns to represent
the scope of the sync operation. The SyncOperation entity was
modified to use new AttributeConverters to hide the conversion
to/from Json.
    
When requested, the FMS will store the SyncOperation in the db
then check whether to accept the request or not based on other
active requests in the db.
    
The SyncOperationAlreadyRunningException was created and mapped
to 409 CONFLICT.

### CODE REVIEW

File: freeipa/src/main/java/com/sequenceiq/freeipa/entity/util/ListToString.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+import java.io.IOException;
+import java.util.List;
+
+import javax.persistence.AttributeConverter;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.sequenceiq.cloudbreak.common.json.Json;
+
+public abstract class ListToString<U> implements AttributeConverter<List<U>, String> {
```
Comment (lacikaaa) [inline freeipa/src/main/java/com/sequenceiq/freeipa/entity/util/ListToString.java]: if you would create sth like `JsonStringSetUtils` then you could use `JsonToString` to store the data. See it in action here: https://github.com/hortonworks/cloudbreak/blob/95e506c796c1b337bc9d7bff945c9a3476fc3c42/core-model/src/main/java/com/sequenceiq/cloudbreak/domain/stack/cluster/DatalakeResources.java#L50-L52
But I can accept your solution if you prefer this way

Comment (handavid) [reply freeipa/src/main/java/com/sequenceiq/freeipa/entity/util/ListToString.java]: I did see that, but I prefer to have the Json hidden from the reset of the application. Only the entity needs to know that the field is stored as text. Also, I'd rather have any deserialization exceptions get thrown when loading the entity to ensure that we always have good data in the table. We may miss problems if we deserialize it lazily (and repetitively)

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/CleanupSyncOperation.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            cleanupStaleSyncOperation(System.currentTimeMillis() - SYNC_OPERATION_TIMEOUT);
+        } catch (Exception e) {
+            LOGGER.error("Failed to clean up SyncOperation table", e);
+        }
+    }
+
+    public void cleanupStaleSyncOperation(Long startedBeforeTime) {
+        List<SyncOperation> staleOperations = syncOperationRepository.findStaleRunning(startedBeforeTime);
+
+        if (!staleOperations.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (handavid) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/CleanupSyncOperation.java]: done

File: freeipa/src/main/resources/schema/app/20190715135810_add_version_and_filters_to_syncoperation.sql [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+UPDATE syncoperation SET version = 1 WHERE version IS NULL;
+UPDATE syncoperation SET environmentlist = '[]' WHERE environmentlist IS NULL;
+UPDATE syncoperation SET userlist = '[]' WHERE userlist IS NULL;
+
+CREATE INDEX IF NOT EXISTS syncoperation_accountid_syncoperationtype_endtime_idx
+  on syncoperation (accountid, syncoperationtype, endtime)
+
+-- //@UNDO
+
+ALTER TABLE syncoperation DROP COLUMN IF EXISTS version;
```
[the remark this answers is not in the document]
Comment (handavid) [reply freeipa/src/main/resources/schema/app/20190715135810_add_version_and_filters_to_syncoperation.sql]: done

**Comment by handavid** [comment]
> do we really need this on rc branch? I would prefer to merge this to master instead

I'm ok either way. I'd prefer to get this in earlier to prevent folks from spamming the sync api by accident. but we currently don't have any protection against it so I guess we can wait.
I'll re-open this against master with your requested changes

**Comment by handavid** [comment]
closing to reopen against master

**Comment by lajosrodek** [comment]
Follow-up to #5652.

**Pull request closed without merging**